keps/sig-instrumentation/1661-event-api-migration/README.md

```diff
@@ -0,0 +1,422 @@
+- For all active series every 30 minutes EventRecorder will create a "heartbeat" call. Goal of this update is to periodically update user on number of occurrences and prevent garbage collection in etcd. The heartbeat will be an Event update that updates the count and last observed time fields in the series field.
+- For all active series every 6 minutes (longer that the longest backoff period) EventRecorder will check if it noticed any attempts to emit isomorphic Event. If there were, it'll check again after aforementioned period (6 minutes). If there weren't it assumes that series is finished and emits closing Event call. This updates the Event by setting state to EventSeriesStateFinished to true and updating the count and last observed time fields in the series field.
+
+##### Short Examples:
+
+After first occurrence, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+}
+```
+After second occurrence, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+  series: {count: 2, state: "Ongoing"},
+}
+```
+After half an hour of crashlooping, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+  series: {count: 4242, state: "Ongoing"},
+}
+```
+Minute after crashloop stopped, Event looks like:
+```
+{
+  regarding: A,
+  action: B,
+  reportingController: C,
+  ...,
+  series: {count: 424242, state: "Finished"},
+}
+```
+
+#### Client Side Changes
+
+All clients will need to eventually migrate to use new Events, but no other actions are required from them. Deduplication logic change will be completely transparent after the move to the new API.
+
+#### Restarts
+
+Event recorder library will list all Events emitted by corresponding components and reconstruct internal activeSeries map from it.
+
+### Defence in Depth
+
+Because Events proved problematic we want to add multiple levels of protection in the client library to reduce chances that Events will be overloading API servers in the future. We propose to do two things.
+
+#### Aggressive Backoff
+
+We need to make sure that kubernetes client used by EventRecorder uses properly configured and backoff pretty aggressively. Events should not destabilize the cluster, so if EventRecorder receives 429 response it should exponentially back off for non-negligible amount of time, to let API server recover.
+
+### Other Related Changes
+
+To allow easier querying we need to make following fields selectable for Events:
+- event.reportingComponent
+- event.reportingInstance
+- event.action
+- event.reason
+- event.regarding...
+- event.related...
+- event.type
+
+Kubectl will need to be updated to use new Events if present.
+
+### Test Plan
+
+Correctness:
+
+- Update all unit tests that are using Event lib to use the new API and make sure they all pass.
+- Run tests with both healthy and crash-looping clusters that keep generating Events to ensure they are produced in an expected way.
```

Comment:
What kind of tests? Manual?

Reply:
I'm thinking of mainly doing tests manually, e.g., create a cluster generating events and check how that works. I'm not familiar with testing framework in Kubernetes. What do you think? Do you have any good suggestions regarding testing?

Reply:
@wojtek-t - do we have any scale tests we can leverage ? (e.g. re-use a job where we pull images, run them then then let containers crashloop)

Reply:
Updated this section.